weld/codegen/llvm2/cmp.rs

```diff
@@ -0,0 +1,351 @@
+                    let cond = LLVMBuildICmp(builder, LLVMIntEQ, field_result, self.i32(0), c_str!(""));
+
+                    // Continue to next only if the field result returns 0 (fields are equal).
+                    LLVMBuildCondBr(builder, cond, next_block, done_block);
+                    
+                    // If field result is nonzero, return field result.
+                    LLVMPositionBuilderAtEnd(builder, done_block);
+                    LLVMBuildRet(builder, field_result);
+                    
+                    LLVMPositionBuilderAtEnd(builder, next_block);
+                }
+                
+                // All blocks equal, return 0.
+                LLVMBuildRet(builder, result); 
+            }
+            // Vectors comprised of unsigned chars or booleans can be compared for using `memcmp`.
+            Vector(ref elem) if elem.supports_memcmp() => {
+                let left_vector = self.load(builder, left)?;
+                let right_vector = self.load(builder, right)?;
+                let left_size = self.gen_size(builder, ty, left_vector)?;
+                let right_size = self.gen_size(builder, ty, right_vector)?;
+
+                // memcmp will run off the end of the smaller buffer,
+                // so emulate strcmp semantics by stopping at the end of the smaller buffer.
+                // Note that this only works when both vectors have elements of the same type.
+                let min_size = gen_binop(builder, Min, left_size, right_size, &Scalar(I64))?;
+
+                let zero = self.i64(0);
+                let left_data = self.gen_at(builder, ty, left_vector, zero)?;
+                let right_data = self.gen_at(builder, ty, right_vector, zero)?;
+                let left_data = LLVMBuildBitCast(builder, left_data, self.void_pointer_type(), c_str!(""));
+                let right_data = LLVMBuildBitCast(builder, right_data, self.void_pointer_type(), c_str!(""));
+                let elem_ty = self.llvm_type(elem)?;
+                let elem_size = self.size_of(elem_ty);
+                let bytes = LLVMBuildNSWMul(builder, min_size, elem_size, c_str!(""));
+
+                // Call memcmp
+                let name = "memcmp";
+                let ret_ty = self.i32_type();
+                let ref mut arg_tys = [self.void_pointer_type(), self.void_pointer_type(), self.i64_type()];
+                if self.intrinsics.add(name, ret_ty, arg_tys) {
+                    let memcmp = self.intrinsics.get(name).unwrap();
+                    LLVMExtAddAttrsOnParameter(self.context, memcmp, &[ReadOnly, NoCapture], 0);
+                    LLVMExtAddAttrsOnParameter(self.context, memcmp, &[ReadOnly, NoCapture], 1);
+                }
+                let ref mut args = [left_data, right_data, bytes];
+                let memcmp_result = self.intrinsics.call(builder, name, args)?;
```

Comment:
This is a little weird because it'll return things like `'a' == 'abc'`, which we don't want. I think if the bytes are equal, you need to compare the sizes to decide the final result. Adding this code (+/- typos) should fix it:

```rust
let func = self.gen_cmp_fn(&Scalar(I64))?;
let mut args = [left_size, right_size];
let size_eq = LLVMBuildCall(builder, func, args.as_mut_ptr(), args.len() as u32, c_str!(""));

let bytes_equal = LLVMBuildICmp(builder, LLVMIntEQ, memcmp_result, self.i32(0), c_str!(""));
let result = LLVMBuildSelect(builder, bytes_equal, size_eq, memcmp_result, c_str!(""));
LLVMBuildRet(builder, result);
```

You should add a test that compares vectors like [1] and [1, 2, 3, 4], where one is a substring of another.